apiserver/facades/client/highavailability/highavailability.go

```diff
@@ -187,31 +220,49 @@ func validateCurrentControllers(st *state.State, cfg controller.Config, machineI
 	return nil
 }
 
-// getReferenceController looks up the ideal controller to use as a reference for Constraints and Series
-func getReferenceController(st *state.State, machineIds []string) (*state.Machine, error) {
-	// Sort the controller IDs from low to high and take the first.
-	// This will typically give the initial bootstrap machine.
-	var controllerIds []int
-	for _, id := range machineIds {
-		idNum, err := strconv.Atoi(id)
+// validatePlacementForSpaces checks whether there are both space constraints
+// and machine placement directives.
+// If there are, checks are made to ensure that the machines specified have at
+// least one address in all of the spaces.
+func validatePlacementForSpaces(st *state.State, spaces *[]string, placement []string) error {
+	if spaces == nil || len(*spaces) == 0 || len(placement) == 0 {
+		return nil
+	}
+
+	for _, v := range placement {
+		p, err := instance.ParsePlacement(v)
 		if err != nil {
-			logger.Warningf("ignoring non numeric controller id %v", id)
+			return errors.Annotate(err, "parsing placement")
+		}
+		if p.Directive == "" {
 			continue
 		}
-		controllerIds = append(controllerIds, idNum)
-	}
-	if len(controllerIds) == 0 {
-		return nil, errors.Errorf("internal error; failed to find any controllers")
-	}
-	sort.Ints(controllerIds)
-	controllerId := controllerIds[0]
 
-	// Load the controller machine and get its constraints.
-	controller, err := st.Machine(strconv.Itoa(controllerId))
-	if err != nil {
-		return nil, errors.Annotatef(err, "reading controller id %v", controllerId)
+		m, err := st.Machine(p.Directive)
```

Comment:
I believe there are several directives which could be supported that may not just be simple machine integers.
Though I think those would all fall under the "machine NotFound" issue and probably be fine.
Like you tested with "--to valid". and the dummy provider.
I believe on MAAS, you can "juju add-machine MAASNAME" and possibly "juju deploy --to MAASNAME"
which makes me think that "juju enable-ha --to MAASNAME1,MAASNAME2" probably should work. Though I don't think we want to go through the effort of validating that MAASNAME1 actually has the right spaces.
Any chance you could try manual testing?
I think on guimaas it would look like:
juju bootstrap
juju enable-ha --to nuc-3,nuc-4

Sort of thing.
It may be that we really should be passing constraints in along with the MAAS name, so that you don't shoot yourself in the foot with an enable-ha --to MAASNAME that isn't in the right spaces.
Though with that sort of syntax, I'm less worried as long as my work finishes and you can "juju remove-machine --force WRONGONE"

Reply:
Didn't know that. I'll poke it this way presently.